include/exadg/grid/grid_utilities.h

```diff
@@ -467,6 +467,104 @@ create_triangulation_with_multigrid(
+  std::vector<PeriodicFacePairs<dim>> &                            coarse_periodic_face_pairs,
+  GridData const &                                                 data,
+  bool const                                                       amr_preserves_boundary_cells)
+{
+  // For serial or distributed triangulations, deal.II can automatically generate the coarse
+  // triangulations.
+  if(data.triangulation_type == TriangulationType::Serial or
+     data.triangulation_type == TriangulationType::Distributed)
+  {
+    if(data.triangulation_type == TriangulationType::Serial)
+    {
+      AssertThrow(
+        data.element_type == ElementType::Hypercube,
+        dealii::ExcMessage(
+          "The create_geometric_coarsening_sequence function of dealii does currently not support "
+          "simplicial elements."));
+
+      coarse_triangulations_const =
+        dealii::MGTransferGlobalCoarseningTools::create_geometric_coarsening_sequence(
+          fine_triangulation);
+    }
+    else if(data.triangulation_type == TriangulationType::Distributed)
+    {
+      coarse_triangulations_const =
+        dealii::MGTransferGlobalCoarseningTools::create_geometric_coarsening_sequence(
+          fine_triangulation,
+          BalancedGranularityPartitionPolicy<dim>(
+            dealii::Utilities::MPI::n_mpi_processes(fine_triangulation.get_communicator())));
+    }
+    else
+    {
+      AssertThrow(false, dealii::ExcMessage("Invalid parameter triangulation_type."));
+    }
+
+    // deal.II adds the fine triangulation as the last entry of the vector. According to our
+    // convention in ExaDG, we only include the triangulations coarser than the fine level. Hence,
+    // we remove the last entry of the vector.
+    coarse_triangulations_const.pop_back();
+
+    // Update periodic face pairs if existent.
+    if(fine_periodic_face_pairs.size() > 0)
+    {
+      AssertThrow(amr_preserves_boundary_cells,
+                  dealii::ExcMessage(
+                    "Combination of adaptive mesh refinement and periodic face pairs"
+                    " requires boundary cells to be preserved."));
+      coarse_periodic_face_pairs.resize(coarse_triangulations_const.size());
+      for(unsigned int level = 0; level < coarse_periodic_face_pairs.size(); ++level)
+      {
+        coarse_periodic_face_pairs[level] = fine_periodic_face_pairs;
+      }
+    }
+  }
+  else if(data.triangulation_type == TriangulationType::FullyDistributed)
+  {
+    AssertThrow(false,
+                dealii::ExcMessage("Combination of adaptive mesh refinement and "
+                                   "TriangulationType::FullyDistributed not implemented."));
+  }
+  else
+  {
+    AssertThrow(false, dealii::ExcMessage("Invalid parameter triangulation_type."));
+  }
+}
+
+/**
+ * Function to update the grid members after adaptive mesh refinement with Grid interface.
+ */
+template<int dim>
+inline void
+setup_after_coarsening_and_refinement(Grid<dim> &      grid,
```

Comment:
This function only seems to exist because of the if-statement below. I think the code would simplify overall when writing on the calling side

```C++
if(param.involves_h_multigrid()`
  create_coarse_triangulations_after_coarsening_and_refinement(...);
```
and then removing the present function.

For comparison, the function https://github.com/exadg/exadg/blob/b651712b088348f4e6684244a7040e961459f14e/include/exadg/grid/grid_utilities.h#L419 writes the if-statement inside the function, but there we can not pull it outside as we can here.

Reply:
I totally see the point. Motivation was, that I wanted to have a call like `GridUtilities::setup_after_coarsening_and_refinement(grid, ...)` to make clear, that one needs to call this.
Now, we will have the if in all drivers, but I can see your point!
So in the current state, where this is the only thing we need to do with the grid after AMR, I totally agree.
Maybe in the future, we will shift things around again. I will address this in a common commit with the above suggested change to `GridUtilities`, since the Asserts I also mentioned before are then moved. Stay tuned :smile: